### PR TITLE
SNOW-206836 Send telemetry message for failed query.

### PR DESCRIPTION
**Pull request by sfc-gh-mrui**
Refer to Jira description for details.
https://snowflakecomputing.atlassian.net/browse/SNOW-206836

### CODE REVIEW

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 1]
```diff
@@ -188,6 +188,31 @@ object SnowflakeTelemetry {
...
+                                         elapse: Long,
+                                         message: String,
+                                         details: String
+                                         ): Unit = {
+    val metric: ObjectNode = mapper.createObjectNode()
+    metric.put(TelemetryQueryResultFields.SPARK_CONNECTOR_VERSION, Utils.VERSION)
+    metric.put(TelemetryQueryResultFields.OPERATION, operation)
+    metric.put(TelemetryQueryResultFields.QUERY, query)
+    metric.put(TelemetryQueryResultFields.QUERY_ID, queryId)
+    metric.put(TelemetryQueryResultFields.QUERY_RESULT, queryResult)
```
Comment (sfc-gh-ema) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:205]: queryResult and message are independent fields, but it seems that message is only used for exception conditions. Do you ever expect 'success', but also with a populated message field?

File: src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala [hunk 2]
```diff
@@ -468,66 +480,118 @@ private[io] object StageWriter {
...
+        // Only load part of missed files.
+        // Exception is raised for the failure.
+        val secondCopyFileSet: Option[mutable.Set[String]] =
+        if (TestHook.isTestFlagEnabled(
+          TestHookFlag.TH_COPY_INTO_TABLE_MISS_FILES_FAIL)) {
+          Some(missedFileSet.grouped(2).toList.head)
+        } else {
+          Some(missedFileSet)
+        }
+        progressMessage.append(
```
Comment (sfc-gh-ema) [inline src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala:522]: What if we hit an issue that's not caught, i.e., there is a hang, or the process ends unexpectedly? Then we don't get any of these progress messages. Should we think about explicitly sending the messages as the progress is updated?

File: src/it/scala/net/snowflake/spark/snowflake/SnowflakeResultSetRDDSuite.scala [hunk 3]
```diff
@@ -607,6 +607,70 @@ class SnowflakeResultSetRDDSuite extends IntegrationSuiteBase {
     }
   }
 
+  // Negative test for read fail.
```
Comment (sfc-gh-ema) [inline src/it/scala/net/snowflake/spark/snowflake/SnowflakeResultSetRDDSuite.scala:610]: Do these tests actually test any of the telemetry functionality?

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala [hunk 4]
```diff
@@ -711,13 +712,38 @@ private[snowflake] class SnowflakeSQLStatement(
 
   def isEmpty: Boolean = list.isEmpty
 
-  def execute(
-    bindVariableEnabled: Boolean
-  )(implicit conn: Connection): ResultSet =
-    if (bindVariableEnabled) executeWithBindVariable(conn)
-    else executeWithoutBindVaribale(conn)
+  private[snowflake] def executeWithoutThrow(
```
Comment (sfc-gh-ema) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:715]: Why not just send telemetry in `catch` instead of testing for `isDefined` for the third tuple field?

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 1 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:205]: Yes. I intend to make it to be easy to support success query.  In the beginning, the message is named as "spark_query_fail". I change it to "spark_query_result" with the QUERY_RESULT field.
The high level idea is for SC dashboard. If we send success message for read and write. we can monitor each customer to see how many spark job they run, how many fail/succeed, and look into details for the fail reason.
We can discuss in the later meeting.

File: src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala (hunk 2 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala:522]: Good catch. We catch "Throwable", so the possibility not to print the message is rare. But it is possible, for example, the task is running and the user kills it.
I will use a string array for the progress. The messages are logged directly. If telemetry message need to be send, all details information is included.

**Review by sfc-gh-ema — APPROVED** [review]

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala [hunk 5]
```diff
@@ -705,12 +705,38 @@ private[snowflake] class SnowflakeSQLStatement(
...
+  : (ResultSet, String, Option[Throwable]) =
+    if (bindVariableEnabled) {
+      executeWithBindVariable(conn)
+    } else {
+      executeWithoutBindVariable(conn)
+    }
 
-  private def executeWithoutBindVariable(conn: Connection): ResultSet = {
+  private def executeWithoutBindVariable(conn: Connection)
+  : (ResultSet, String, Option[Throwable]) = {
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:739]: Create a case class for this return type.

Comment (sfc-gh-sshankar) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:739]: Also, this is not a good way to write this.
```
case class ResultInfo(queryID: String, result: Option[ResultSet])
private def executeWithoutBindVariable(conn): ResultInfo
```
And use Try/Success/Failure to deal with exceptions.
```
def execute() = {
  val executeResult = Try(...)
}
```

Comment (sfc-gh-sshankar) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:739]: Or better yet, just use try/catch.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 6]
```diff
@@ -188,6 +188,31 @@ object SnowflakeTelemetry {
     )
   }
 
+  private[snowflake] def sendQueryResult(conn: Connection,
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:191]: is this sending the query result or the query status ?

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 7]
```diff
@@ -198,9 +223,36 @@ object TelemetryTypes extends Enumeration {
...
+  // Spark connector version
+  val SPARK_CONNECTOR_VERSION: String = "spark_connector_version"
+  // The unsupported operation for pushdown
+  val OPERATION: String = "operation"
+  // Query statement
+  val QUERY: String = "query"
+  // query ID if available
+  val QUERY_ID: String = "queryid"
+  // Query result: fail vs success
+  val QUERY_RESULT: String = "result"
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:247]: result != status

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 8]
```diff
@@ -198,9 +223,36 @@ object TelemetryTypes extends Enumeration {
...
+  // query execution time
+  val QUERY_ELAPSE: String = "elapse"
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:249]: elapsed_time

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 9]
```diff
@@ -198,9 +223,36 @@ object TelemetryTypes extends Enumeration {
...
+  // The error message for the exception
+  val EXCEPTION_MESSAGE: String = "message"
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:251]: exception_message

File: src/main/scala/net/snowflake/spark/snowflake/io/StageReader.scala [hunk 10]
```diff
@@ -36,13 +36,31 @@ private[snowflake] object StageReader {
...
       statement,
       s"@$stage/$prefix/",
       compressFormat,
       format
-    ).execute(params.bindVariableEnabled)(conn)
+    )
+
+    val startTime = System.currentTimeMillis()
+    val (res, queryID, errorThrow) =
+      copyStatement.executeWithoutThrow(params.bindVariableEnabled)(conn)
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/io/StageReader.scala:49]: Why do we need this, aren't we logging telemetry in execute ?

File: src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala [hunk 11]
```diff
@@ -468,66 +482,115 @@ private[io] object StageWriter {
...
-      )
-
-      def getMissedFileInfo(missedFileSet: mutable.Set[String]): String = {
-        s"""missedFileCount=${missedFileSet.size}
-           | Files: (${missedFileSet.mkString(", ")})
-           |""".stripMargin.filter(_ >= ' ')
+    var lastQueryID: String = null
+    try {
+      // execute the COPY INTO TABLE statement
+      val (resultSet, queryID, errorThrow) =
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala:491]: This is super-ugly, you shouldn't have to change this stuff,

File: src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala [hunk 12]
```diff
@@ -468,66 +482,115 @@ private[io] object StageWriter {
...
+      // Get missed files if there are any.
+      var missedFileSet = getCopyMissedFiles(params, resultSet, expectedFileSet)
 
-      // It is expected all the files must be loaded.
+      // If any files are missed, execute 2nd COPY command
       if (missedFileSet.nonEmpty) {
-        throw new SnowflakeConnectorException(
-          s"""These files are missed when COPY INTO TABLE:
-             | ${getMissedFileInfo(missedFileSet)}
+        // Negative test:
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala:513]: This seems unrelated.

**Review by sfc-gh-sshankar — COMMENTED** [review]
I think this needs more work.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala (hunk 5 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:739]: The reason to work in this way to retrieve the queryID for query fail.
For successful query, we can get the queryID from resultSet. For failed query, we have to get the queryID from the Statement. But the statement is created in side of executeWithoutBindVariable()/executeWithBindVariable().
To get the queryID and not to introduce duplicate code, I change the internal function to return the Exception and the caller processes the exception.

But I agree it is good idea to use CASE CLASS for the return type. I plan to use ExecuteResult.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 6 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:191]: query status. I will change the function name as sendQueryStatus()

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 8 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:249]: will update

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 7 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:247]: Will change:
TelemetryQueryResultFields  -> TelemetryQueryStatusFields
QUERY_RESULT -> QUERY_STATUS

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 9 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:251]: will update

File: src/main/scala/net/snowflake/spark/snowflake/io/StageReader.scala (hunk 10 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/io/StageReader.scala:49]: executeWithoutThrow() doesn't log error inside.
This is to sending a telemetry message when reading with COPY INTO LOCATION fails.
The telemetry message is independent to send log entries to telemetry.

PS: I am a little confusing about the question. If this doesn't address your comment, could you please explain the question?

File: src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala (hunk 11 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala:491]: It's some complex, but I think it is very useful for diagnosis.
I can just change to call executeWithoutThrow(), check error and send telemetry message.
That will be easier, but I think the current is better.
With current change, if COPY fails, we can get details info about the progress and situation.

File: src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala (hunk 12 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/io/StageWriter.scala:513]: This is not new code. I guess it makes confused because the format change causes it to be like new code.

File: src/main/scala/net/snowflake/spark/snowflake/io/StageReader.scala (hunk 10 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/io/StageReader.scala:49]: Oh, I see. I think you mean we can move to send the telemetry message inside of executeWithoutThrow().
It's doable, I didn't do that because some information for the telemetry message are collected by the caller.
If we want to send message inside of executeWithoutThrow(), we may add some extra parameters.
In the future, we will send successful messages too, some info are not ready in the executeWithoutThrow() function such as total data size, partition count, etc.
So, I think it should be better to process it by caller and keep executeWithoutThrow() simple.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 9 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:258]: After considering it again, I change it back to "message".
The reason is that we have already sent Exception with OOB telemetry message on executors. I think we'd better  use the same field name for exception. Otherwise, it may confusing when querying the message from snowhouse.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala [hunk 13]
```diff
@@ -691,6 +691,10 @@ private[snowflake] class SnowflakeSQLStatement(
 
   private val log = LoggerFactory.getLogger(getClass)
 
+  private var lastQueryID: String = _
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:694]: This should be a thread local, no ?

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 14]
```diff
@@ -82,7 +82,7 @@ object SnowflakeTelemetry {
                        success: Boolean,
                        useProxy: Boolean,
                        queryID: Option[String],
-                       exception: Option[Exception]): Unit =
+                       errorThrow: Option[Throwable]): Unit =
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:85]: errorThrow => throwable

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 15]
```diff
@@ -198,9 +230,40 @@ object TelemetryTypes extends Enumeration {
...
+
+object TelemetryQueryStatusFields {
+  // Spark connector version
+  val SPARK_CONNECTOR_VERSION: String = "spark_connector_version"
+  // The query operation: read vs write
+  val OPERATION: String = "operation"
+  // Query statement
+  val QUERY: String = "query"
+  // query ID if available
+  val QUERY_ID: String = "queryid"
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:252]: query_id. Please have consistent style

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala (hunk 13 above)
Comment (sfc-gh-sshankar) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:694]: This is only set on failure. I think it would be useful to always set it.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 14 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:85]: will update

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala (hunk 13 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:694]: yes. It has set up for both failure and success. For example,
      val rs = DefaultJDBCWrapper.executePreparedQueryInterruptibly(statement)
      lastQueryID = rs.asInstanceOf[SnowflakeResultSet].getQueryID
      rs

Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeJDBCWrapper.scala:694]: SnowflakeSQLStatement is defined as "class SnowflakeSQLStatement", not "object SnowflakeSQLStatement". The caller doesn't share SnowflakeSQLStatement instance between thread. So it is not necessary to make it thread local.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala (hunk 15 above)
Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:252]: I will use QUERYID to align OOB message.

File: src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala [hunk 16]
```diff
@@ -249,7 +249,7 @@ object TelemetryQueryStatusFields {
   // Query statement
   val QUERY: String = "query"
   // query ID if available
-  val QUERY_ID: String = "queryid"
+  val QUERYID: String = "queryid"
```
Comment (sfc-gh-sshankar) [inline src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala:252]: query_id, to match max_retry and use_proxy, etc.

Comment (sfc-gh-mrui) [reply src/main/scala/net/snowflake/spark/snowflake/SnowflakeTelemetry.scala]: Will update

**Review by sfc-gh-sshankar — APPROVED** [review]
